sdk-base/src/main/java/com/mapbox/maps/extension/observable/ObservableExtension.kt

```diff
@@ -301,124 +298,282 @@ fun ObservableInterface.unsubscribeSourceRemoved(observer: Observer) =
  * Get the parsed event data for resource-request event.
  * @return a parsed ResourceEventData object.
  */
+@Suppress("UNCHECKED_CAST")
 fun Event.getResourceEventData(): ResourceEventData {
-  val json = data.toJson()
-  return gson.fromJson(json, ResourceEventData::class.java)
+  val map = data.contents as Map<String, Value>
+  val requestMap = map[REQUEST]!!.contents as Map<String, Value>
+  val responseMap = map[RESPONSE]?.contents as Map<String, Value>?
+  val errorMap = requestMap[ERROR]?.contents as Map<String, Value>?
+  return ResourceEventData(
+    begin = map[BEGIN]!!.contents as Long,
+    end = map[END]?.contents as Long?,
+    dataSource = DataSourceType.valueOf(map[DATA_SOURCE]!!.contents as String),
+    request = Request(
+      loadingMethod = requestMap[LOADING_METHOD]!!.contents as List<String>,
+      url = requestMap[URL]!!.contents as String,
+      kind = RequestType.valueOf(requestMap[KIND]!!.contents as String),
+      priority = RequestPriority.valueOf(requestMap[PRIORITY]!!.contents as String)
+    ),
+    response = if (responseMap == null) null else Response(
+      eTag = responseMap[E_TAG]?.contents as String?,
+      mustRevalidate = responseMap[MUST_REVALIDATE]!!.contents as Boolean,
+      noContent = responseMap[NO_CONTENT]!!.contents as Boolean,
+      modified = responseMap[MODIFIED]?.contents as String?,
+      source = ResponseSourceType.valueOf(responseMap[SOURCE]!!.contents as String),
+      notModified = responseMap[NOT_MODIFIED]!!.contents as Boolean,
+      expires = responseMap[EXPIRES]?.contents as String?,
+      size = responseMap[SIZE]!!.contents as Int,
+      error = if (errorMap == null) null else Error(
+        reason = ResponseErrorReason.valueOf(errorMap[REASON]!!.contents as String),
+        message = errorMap[MESSAGE]!!.contents as String
+      )
+    ),
+    cancelled = map[CANCELLED]!!.contents as Boolean
+  )
 }
 
 /**
  * Get the parsed event data for map loaded event.
  * @return a parsed MapLoadedEventData object.
  */
+@Suppress("UNCHECKED_CAST")
 fun Event.getMapLoadedEventData(): MapLoadedEventData {
-  val json = data.toJson()
-  return gson.fromJson(json, MapLoadedEventData::class.java)
+  val map = data.contents as Map<String, Value>
+  return MapLoadedEventData(
+    begin = map[BEGIN]!!.contents as Long,
+    end = map[END]?.contents as Long?
+  )
 }
 
 /**
  * Get the parsed event data for map loading error event.
  * @return a parsed MapLoadingErrorEventData object.
  */
+@Suppress("UNCHECKED_CAST")
 fun Event.getMapLoadingErrorEventData(): MapLoadingErrorEventData {
-  val json = data.toJson()
-  return gson.fromJson(json, MapLoadingErrorEventData::class.java)
+  val map = data.contents as Map<String, Value>
+  val tileIDMap = map[TILE_ID]?.contents as Map<String, Value>?
+  return MapLoadingErrorEventData(
+    begin = map[BEGIN]!!.contents as Long,
```

Comment:
Let's create several extensions:
```
// or Map<String, Value>.requiredLong(name:String) {...}

fun Map<String, Value>.nonNullLong(name: String):Long{
   return this[name]!!.contents as Long
}
fun Map<String, Value>.nullableLong(name: String):Long?{
   return this[name]?.contents as Long?
}
```

same for `String`, `Int`, `Double` and maybe one for custom type
it will be more readable

Reply:
Nice suggestion.